apollo-router/src/services/execution_service.rs

```diff
@@ -80,6 +88,67 @@ where
     }
 }
 
+// modifies the response stream to set `has_next` to `false` on the last response
+fn filter_stream(first: Response, mut stream: Receiver<Response>) -> Receiver<Response> {
+    let (mut sender, receiver) = futures::channel::mpsc::channel(10);
+
+    tokio::task::spawn(async move {
+        consume_responses(first, &mut stream, &mut sender).await?;
+
+        while let Some(current_response) = stream.next().await {
+            consume_responses(current_response, &mut stream, &mut sender).await?;
+        }
+        println!("done");
+        Ok::<_, SendError>(())
+    });
+
+    receiver
+}
+
+async fn consume_responses(
```

Comment:
I think there may be a race condition here.

In the case where `stream.try_next()` is called and returns an error as there may be more items in the stream if the stream is then closed before the call to `next` in `filter_stream` there will be no final empty response with `has_next: false`

Reply:
`try_next` is like doing `next` but without await: if there's a message in the stream, it will be returned by `try_next`, if not then `try_next` will return an error and the next call to `next` would wait for it.

In the case you are describing, `try_next` would not return an error if there are in flight messages. In the case where `try_next ` would return an error, then somehow between returning from `consume_responses` and the call to `next` the stream gets new messages then is closed, then `next ` would return a message, some calls to `try_next` would return messages, then when there's nothing remaining  `try_next` would return Ok(None)`.

The one possible race I worry about is if messages are received and re-sent, then we await on `next`, then for whatever reason the stream is disconnected (maybe all the senders are dropped). Then we would need to add a final `has_next = false` response. But I don't see how this could play out

Reply:
@BrynCooke that last cvase should be addressed by https://github.com/apollographql/router/pull/1745/commits/5210765bbf723d2e9b20327b42742157e15fb67d

Reply:
LGTM!